File: src/main/java/com/kinisoftware/upcomingMovies/handler/NewReleasesIntentHandler.java
package com.kinisoftware.upcomingMovies.handler;

import java.time.LocalDate;
import java.time.temporal.WeekFields;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static com.kinisoftware.upcomingMovies.UpcomingMoviesStreamHandler.CARD_TITLE;

import com.amazon.ask.dispatcher.request.handler.HandlerInput;
import com.amazon.ask.dispatcher.request.handler.RequestHandler;
import com.amazon.ask.model.Intent;
import com.amazon.ask.model.IntentRequest;
import com.amazon.ask.model.Request;
import com.amazon.ask.model.Response;
import com.amazon.ask.model.Slot;
import com.amazon.ask.request.Predicates;
import com.amazon.ask.response.ResponseBuilder;
import com.kinisoftware.upcomingMovies.ReleasesGetter;
import com.kinisoftware.upcomingMovies.model.NewRelease;

public class NewReleasesIntentHandler implements RequestHandler {

	@Override
	public boolean canHandle(HandlerInput input) {
		return input.matches(Predicates.intentName("NewReleasesIntent"));
	}

	@Override
	public Optional<Response> handle(HandlerInput input) {
		Request request = input.getRequestEnvelope().getRequest();
		IntentRequest intentRequest = (IntentRequest) request;
		Intent intent = intentRequest.getIntent();
		Map<String, Slot> slots = intent.getSlots();

		Slot releasesDate = slots.get("releasesDate");
		String dateValue = releasesDate.getValue();
		System.out.println("Slot value:" + dateValue);

		String text = null;
		String reprompText = null;
		if (dateValue == null) {
			reprompText = "Lo siento, no he entendido la fecha que quieres consultar. ¿Podrías repetir, por favor?";
		} else {
			List<String> movies = ReleasesGetter.get()
					.parallelStream()
					.filter(isReleasedOnDate(dateValue))
					.map(NewRelease::getTitle)
					.collect(Collectors.toList());

			if (movies.isEmpty()) {
				text = "Lo siento, aún no tengo estrenos para esa fecha";
			} else {
				text = "Las películas que se estrenan son: ".concat(String.join(", ", movies));
			}
		}
		System.out.println("Output:" + text);

		ResponseBuilder responseBuilder = input.getResponseBuilder().withSpeech(text).withSimpleCard(CARD_TITLE, text);
		if (reprompText != null) {
			responseBuilder.withReprompt(reprompText);
		}

		return responseBuilder.build();
	}

	private Predicate<NewRelease> isReleasedOnDate(String dateValue) {
		return newRelease -> getWeekFormatStyle(newRelease.getReleaseDate()).equals(dateValue) || getMonthFormatStyle(newRelease.getReleaseDate()).equals(dateValue);
	}

	private String getWeekFormatStyle(String releaseDate) {
		LocalDate localDate = LocalDate.parse(releaseDate);
		WeekFields weekFields = WeekFields.of(Locale.getDefault());
		return localDate.getYear() + "-W" + localDate.get(weekFields.weekOfWeekBasedYear());
	}

	private String getMonthFormatStyle(String releaseDate) {
		LocalDate localDate = LocalDate.parse(releaseDate);
		return localDate.getMonthValue() < 10
				? localDate.getYear() + "-0" + localDate.getMonthValue()
				: localDate.getYear() + "-" + localDate.getMonthValue();
	}
}

File: src/main/kotlin/com/kinisoftware/upcomingMovies/handler/NewReleasesIntentHandler.kt
package com.kinisoftware.upcomingMovies.handler

import com.amazon.ask.dispatcher.request.handler.HandlerInput
import com.amazon.ask.dispatcher.request.handler.RequestHandler
import com.amazon.ask.exception.AskSdkException
import com.amazon.ask.model.IntentRequest
import com.amazon.ask.model.Response
import com.amazon.ask.model.interfaces.alexa.presentation.apl.RenderDocumentDirective
import com.amazon.ask.request.Predicates
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.kinisoftware.upcomingMovies.MoviesGetter
import com.kinisoftware.upcomingMovies.Translations
import com.kinisoftware.upcomingMovies.getLanguage
import com.kinisoftware.upcomingMovies.getResponse
import com.kinisoftware.upcomingMovies.getTitle
import com.kinisoftware.upcomingMovies.supportAPL
import java.io.File
import java.io.IOException
import java.util.HashMap
import java.util.Optional

class NewReleasesIntentHandler(private val moviesGetter: MoviesGetter) : RequestHandler {

    override fun canHandle(input: HandlerInput): Boolean {
        return input.matches(Predicates.intentName("NewReleasesIntent"))
    }

    override fun handle(input: HandlerInput): Optional<Response> {
        val request = input.requestEnvelope.request
        val intentRequest = request as IntentRequest
        val intent = intentRequest.intent
        val slots = intent.slots
        val releasesDate = slots["releasesDate"]!!
        val dateValue = releasesDate.value
        val locale = intentRequest.locale
        val language = input.getLanguage()

        val movies = moviesGetter.getUpcomings(locale, dateValue)

        return when {
            input.supportAPL() && movies.isNotEmpty() -> {
                try {
                    val mapper = ObjectMapper()
                    val documentMapType = object : TypeReference<HashMap<String, Any>>() {}
                    val document = mapper.readValue<Map<String, Any>>(File("upcomingMoviesScreen.json"), documentMapType)
                    val dataSourceMapType = object : TypeReference<HashMap<String, Any>>() {}
                    val dataSource = mapper.readValue<Map<String, Any>>(File("upcomingMoviesScreenData.json"), dataSourceMapType)

                    val newReleases = mapOf("newReleases" to movies)
                    val documentDirective = RenderDocumentDirective.builder()
                            .withToken("newReleasesSkillAPLToken")
                            .withDocument(document)
                            .withDatasources(dataSource)
                            .putDatasourcesItem("movies", newReleases)
                            .build()

                    return input.responseBuilder
                            .withSpeech("Te muestro los próximos estrenos de cine")
                            .addDirective(documentDirective)
                            .build()
                } catch (e: IOException) {
                    throw AskSdkException("Unable to read or deserialize upcoming movies data", e)
                }
            }
            else -> {
                var shouldEndSession: Boolean
                var text: String
                if (movies.isEmpty()) {
                    text = Translations.getMessage(language, Translations.TranslationKey.UPCOMINGS_NOT_FOUND) +
                            Translations.getMessage(language, Translations.TranslationKey.ASKING_FOR_NOW_PLAYING)
                    shouldEndSession = false
                } else {
                    text = "${Translations.getMessage(language, Translations.TranslationKey.UPCOMINGS_RESPONSE)}: " +
                            movies.map { it.getTitle() }.getResponse(language)
                    shouldEndSession = true
                }
                input.responseBuilder
                        .withSpeech(text)
                        .withShouldEndSession(shouldEndSession)
                        .build()
            }
        }
    }
}
